modules/tensor_mechanics/src/materials/lagrangian/ComputeSimoHughesJ2PlasticityStress.C

```diff
@@ -0,0 +1,195 @@
+    _elasticity_tensor(getMaterialProperty<RankFourTensor>(_elasticity_tensor_name)),
+    _F_old(getMaterialPropertyOld<RankTwoTensor>(_base_name + "deformation_gradient")),
+    _ep_name(_base_name + "effective_plastic_strain"),
+    _ep(declareProperty<Real>(_ep_name)),
+    _ep_old(getMaterialPropertyOldByName<Real>(_ep_name)),
+    _be(declareProperty<RankTwoTensor>(_base_name +
+                                       "volume_preserving_elastic_left_cauchy_green_strain")),
+    _be_old(getMaterialPropertyOldByName<RankTwoTensor>(
+        _base_name + "volume_preserving_elastic_left_cauchy_green_strain")),
+    _Np(declareProperty<RankTwoTensor>(_base_name + "flow_direction")),
+    _flow_stress_material(nullptr),
+    _flow_stress_name(_base_name + "flow_stress"),
+    _H(getMaterialPropertyByName<Real>(_flow_stress_name)),
+    _dH(getDefaultMaterialPropertyByName<Real, false>(
+        derivativePropertyName(_flow_stress_name, {_ep_name}))),
+    _d2H(getDefaultMaterialPropertyByName<Real, false>(
+        derivativePropertyName(_flow_stress_name, {_ep_name, _ep_name})))
+{
+}
+
+void
+ComputeSimoHughesJ2PlasticityStress::initialSetup()
+{
+  _flow_stress_material = &getMaterial("flow_stress_material");
+
+  // Enforce isotropic elastic tensor
+  if (!hasGuaranteedMaterialProperty(_elasticity_tensor_name, Guarantee::ISOTROPIC))
+    mooseError("ComputeSimoHughesJ2PlasticityStress requires an isotropic elasticity tensor");
+}
+
+void
+ComputeSimoHughesJ2PlasticityStress::initQpStatefulProperties()
+{
+  ComputeLagrangianStressPK1::initQpStatefulProperties();
+  _be[_qp].setToIdentity();
+  _ep[_qp] = 0;
+}
+
+void
+ComputeSimoHughesJ2PlasticityStress::computeQpPK1Stress()
+{
+  usingTensorIndices(i, j, k, l, m);
+  const Real G = ElasticityTensorTools::getIsotropicShearModulus(_elasticity_tensor[_qp]);
+  const Real K = ElasticityTensorTools::getIsotropicBulkModulus(_elasticity_tensor[_qp]);
+  const auto I = RankTwoTensor::Identity();
+  const auto Fit = _F[_qp].inverse().transpose();
+  const auto detJ = _F[_qp].det();
+
+  // Update configuration
+  RankTwoTensor f = _inv_df[_qp].inverse();
+  RankTwoTensor f_bar = f / std::cbrt(f.det());
+
+  // Elastic predictor
+  _be[_qp] = f_bar * _be_old[_qp] * f_bar.transpose();
+  RankTwoTensor s = G * _be[_qp].deviatoric();
+  _Np[_qp] = MooseUtils::absoluteFuzzyEqual(s.norm(), 0) ? std::sqrt(1. / 2.) * I
+                                                         : std::sqrt(3. / 2.) * s / s.norm();
+  Real s_eff = s.doubleContraction(_Np[_qp]);
+
+  // Compute the derivative of the strain before return mapping
+  if (_fe_problem.currentlyComputingJacobian())
+    _d_be_d_F = _F_old[_qp].inverse().times<l, m, i, j, k, m>(
+        (I.times<i, k, j, l>(f_bar * _be_old[_qp].transpose()) +
+         I.times<j, k, i, l>(f_bar * _be_old[_qp])) /
+            std::cbrt(f.det()) -
+        2. / 3. * _be[_qp].times<i, j, l, k>(_inv_df[_qp]));
```

Comment:
These loops and others are hard to follow, probably because they have to.  Is there any algorithm in the literature you can point to in a comment?

Reply:
Sorry which loops are you referring to? If you mean the overall algorithm, then I have included a writeup in the PR description. I think there is also a paper by Simo which has this algorithm. I'll find it and cite it in the documentation.

Reply:
I ended up not selecting any loop... Yeah, I was referring to the overall algorithm. My bad, I missed that document you attached to the description: Can that document be md-fied? I am afraid we may not come across it when we or users most need it.

Reply:
Yep, the algorithm is included in the md file.